use AWS price index for Network Loadbalancer hourly costs

skoef opened the pull request:
## What does this PR change?
* Instead of using a fixed price for hourly usage of AWS load balancers, use the price index to detect the correct price for that region

## Does this PR relate to any other PRs?
* Not really but I would like to backport #2766 into this PR before we merge

## How will this PR impact users?
* They will see slightly different costs for loadbalancer usage, especially when outside us-east-2 region (and perhaps other US regions as well)

## How was this PR tested?
* First of all it fixes the JSON syntax error in the fixture from `Test_populate_pricing` so I had to extend to expected product terms as well in that test. Then I added a product and a term from the actual us-east-2 price index to the fixture and asserted that there was indeed a productTerm for it in the result. I also compared the result of `LoadBalancerPricing()` with the price in the fixture.

## Does this PR require changes to documentation?
* I don't think it does, unless somewhere it's documented that there is a fixed price used for loadbalancers in the AWS provider.

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
* No, don't know how

skoef committed:
use AWS price index for Network Loadbalancer hourly costs

Signed-off-by: Reinier Schoof <reinier@skoef.nl>

thomasvn commented on pkg/cloud/aws/provider.go:
```diff
@@ -1036,6 +1038,19 @@ func (aws *AWS) populatePricing(resp *http.Response, inputkeys map[string]bool)
 					skusToKeys[product.Sku] = key
 					aws.ValidPricingKeys[key] = true
 					aws.ValidPricingKeys[spotKey] = true
+				} else if strings.Contains(product.Attributes.UsageType, "LoadBalancerUsage") && product.Attributes.Operation == "LoadBalancing:Network" {
```
Thoughts on filtering for `product.Attributes.Operation == "LoadBalancing"` instead? According to AWS pricing this would be a standard Elastic Load Balancer.

```json
    "T3YPSF6NJ69JWWNM" : {
      "sku" : "T3YPSF6NJ69JWWNM",
      "productFamily" : "Load Balancer",
      "attributes" : {
        "servicecode" : "AmazonEC2",
        "location" : "US East (N. Virginia)",
        "locationType" : "AWS Region",
        "group" : "ELB:Balancer",
        "groupDescription" : "Standard Elastic Load Balancer",
        "usagetype" : "LoadBalancerUsage",
        "operation" : "LoadBalancing",
        "regionCode" : "us-east-1",
        "servicename" : "Amazon Elastic Compute Cloud"
      }
    },
```

skoef replied:
The current implementation of the costmodel only takes kubernetes Services of type Loadbalancer into account for calculation costs. Those are all Network Load Balancers nowadays. When using AWS's loadbalancer controller you can have it manage your kubernetes Ingresses as well and for that it will create an Application Load Balancers.

There used to be a product called Classic Load Balancer that has been [deprecated](https://aws.amazon.com/blogs/aws/ec2-classic-is-retiring-heres-how-to-prepare/) but I don't know how "Standard Load Balancer" falls within AWS's proposition to be honest.

thomasvn replied:
@skoef Do you happen to have any screenshots showing that LoadBalancers are always of type "Network Load Balancer" as opposed to "Application Load Balancer"? It would be good to attach that screenshot to the PR's description.

Also, confirming what you said. CostModel is setting the `kubecost_load_balancer_cost` metric here ([code ref](https://github.com/opencost/opencost/blob/cec0575fcd4469309b5d69346f623fc8067aab7a/pkg/costmodel/metrics.go#L574-L591)). When calculating the cost for this metric it primarily checks Services of Type "LoadBalancer" ([code ref](https://github.com/opencost/opencost/blob/cec0575fcd4469309b5d69346f623fc8067aab7a/pkg/costmodel/costmodel.go#L1341-L1379)).

skoef replied:
@thomasvn I can't provide a screenshot, but did find in the code for AWS loadbalancer controller that it's hardcoded to always use Loadbalancers of type Network ([code ref](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/pkg/service/model_build_load_balancer.go#L87)) for Service based loadbalancers. So for me it made sense to filter specifically on prices for loadbalancers of type Network. Again, this controller itself is able to manage Application Loadbalancers as well, but those are managed through Ingress and opencost doesn't do any accounting on those (yet).

Looking further through the price specifications, there are 3 flavors of loadbalancers: Network, Application and "Standard", but AWS doesn't publicly offer a "Standard" product, so I assumed that that was about the Classic loadbalancers.

thomasvn replied:
@skoef I just checked my environments and found that my Nginx Ingress Controller creates a Classic/Standard LoadBalancer.

Regardless, I think your current implementation is a good improvement and am ok defaulting to Network LoadBalancers. I think the cost difference is minimal. For example in us-east-1 the hourly rate of a Classic vs Network LoadBalancer is `.025/hr` vs `0.0225/hr` respectively.

It can be confirmed by looking at the following pricing sheet https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/us-east-1/index.json. LoadBalancerNetwork = SKU PQ44QRGZYQ87XPN2. LoadBalancer = SKU T3YPSF6NJ69JWWNM.

skoef commented:
@thomasvn please re-review this

thomasvn commented:
Added a comment!

cliffcolvin commented:
@skoef it looks like @thomasvn responded. Is this still in flight? I'm triaging PR to get eyes and eta on things and I can line up some review and testing.

skoef commented:
@cliffcolvin thank you for bringing this to my attention, I replied!

thomasvn reviewed (approved):
Thanks @skoef for this enhancement! Looks good to me.

skoef commented:
Thanks @thomasvn I will pull the last changes since it has conflicts currently. If it requires additional approval afterwards, I'll ping you!

skoef committed:
Merge branch 'develop' of https://github.com/opencost/opencost into awsLoadbalancerCost

skoef committed:
move AWS fixtures for testing in separate files

This way it is way easier to check if the JSON is still valid and it doesn't
clutter the code so much.

Signed-off-by: Reinier Schoof <reinier@skoef.nl>

skoef commented:
@thomasvn I took the liberty of moving all the fixture JSON to separate files because merge conflicts are quite hard to fix there. During developing this PR I already found that the us-east-2 fixture data had syntax errors in it which weren't covered in the tests at all.

thomasvn commented:
Thanks @skoef, I agree with that approach and it leaves the tests cleaner!

thomasvn commented:
@cliffcolvin Could you help to validate whether we are passing test coverage here? It says that there is only "24.9% Coverage on New Code", however when looking at SonarCloud I see lots of files which were not modified by this PR.

cliffcolvin commented:
> @cliffcolvin Could you help to validate whether we are passing test coverage here? It says that there is only "24.9% Coverage on New Code", however when looking at SonarCloud I see lots of files which were not modified by this PR.

@thomasvn and @skoef will dig in and get an answer asap. Thank you both!

cliffcolvin commented:
Sorry about this, I’ve been heads down for a while. @ameijer would you be willing to take a look here? Trying to determine why coverage gate is failing, but would love some help making sure we can merge this or what advice we may have on coverage.

cliffcolvin committed:
Merge branch 'develop' into awsLoadbalancerCost

cliffcolvin commented:
I've been digging, but we are going to bypass this check for now until I can figure out what is causing this question on coverage. Since Thomas N helped verify and test this I'll get this merged as soon as checks pass after I updated the branch.
Thanks @skoef for the contribution here!!
